libp2p/protocols/secure/noise.nim

```diff
@@ -0,0 +1,575 @@
+import ../../peer
+import ../../peerinfo
+import ../../protobuf/minprotobuf
+import secure,
+       ../../crypto/[crypto, chacha20poly1305, curve25519, hkdf],
+       ../../stream/bufferstream
+
+template unimplemented: untyped =
+  doAssert(false, "Not implemented")
+
+logScope:
+  topic = "Noise"
+
+const
+  # https://godoc.org/github.com/libp2p/go-libp2p-noise#pkg-constants
+  NoiseCodec* = "/noise"
+  
+  PayloadString = "noise-libp2p-static-key:"
+
+  ProtocolXXName = "Noise_XX_25519_ChaChaPoly_SHA256"
+
+  # Empty is a special value which indicates k has not yet been initialized.
+  EmptyKey: ChaChaPolyKey = [0.byte, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]
+  NonceMax = uint64.high - 1 # max is reserved
+                             #
+  NoiseMessageMax = 65535
+
+type
+  KeyPair = object
+    privateKey: Curve25519Key
+    publicKey: Curve25519Key
+  
+  # https://noiseprotocol.org/noise.html#the-cipherstate-object
+  CipherState = object
+    k: ChaChaPolyKey
+    # noise spec says uint64
+    # go implementation has uint32 tho I noticed!!
+    n: uint64
+
+  # https://noiseprotocol.org/noise.html#the-symmetricstate-object
+  SymmetricState = object
+    cs: CipherState
+    ck: ChaChaPolyKey
+    h: MDigest[256]
+
+  # https://noiseprotocol.org/noise.html#the-handshakestate-object
+  HandshakeState = object
+    ss: SymmetricState
+    s: KeyPair
+    e: KeyPair
+    rs: Curve25519Key
+    re: Curve25519Key
+
+  HandshakeResult = object
+    cs1: CipherState
+    cs2: CipherState
+    remoteP2psecret: seq[byte]
+    rs: Curve25519Key
+  
+  Noise* = ref object of Secure
+    localPrivateKey: PrivateKey
+    localPublicKey: PublicKey
+    noisePrivateKey: Curve25519Key
+    noisePublicKey: Curve25519Key
+    commonPrologue: seq[byte]
+
+  NoiseConnection* = ref object of SecureConnection
+    readCs: CipherState
+    writeCs: CipherState
+
+  NoiseHandshakeError* = object of CatchableError
+  NoiseDecryptTagError* = object of CatchableError
+  NoiseOversizedPayloadError* = object of CatchableError
+
+# Utility
+
+proc genKeyPair(): KeyPair =
+  result.privateKey = Curve25519Key.random()
+  result.publicKey = result.privateKey.public()
+    
+# todo this should be exposed maybe or replaced with a more public version
+proc getBytes*(key: string): seq[byte] =
+  result.setLen(key.len)
+  copyMem(addr result[0], unsafeaddr key[0], key.len)
+
+proc hashProtocol(name: string): MDigest[256] =
+  # If protocol_name is less than or equal to HASHLEN bytes in length,
+  # sets h equal to protocol_name with zero bytes appended to make HASHLEN bytes.
+  # Otherwise sets h = HASH(protocol_name).
+
+  if name.len <= 32:
+    copyMem(addr result.data[0], unsafeAddr name[0], name.len)
+  else:
+    result = sha256.digest(name)
+
+proc dh(priv: Curve25519Key, pub: Curve25519Key): Curve25519Key =
+  Curve25519.mul(result, pub, priv)
+
+# Cipherstate
+
+proc init(cs: var CipherState; key: ChaChaPolyKey) =
```

Comment:
so typically, we'd do 

```func init(T: type CipherState, key: ..): T = T(
  k: key,
  n: 0
)
```

which allows calling it with `CipherState.init(key)`, almost like a constructor

Reply:
I'm updating internal states by reference tho, within `SymmetricState`, that's why it's a `var`
Further more this is internal stuff really.
Maybe the `init` name is confusing.